src/Suave.Tests/Owin.fs

```diff
@@ -148,5 +159,190 @@ let endToEnd =
+        eq "Reason Phrase set by server" "No Content" result.ReasonPhrase
+
+        match result.Headers.TryGetValues("X-Custom-Before") with
+        | true, actual ->
+          eqs "Headers set after the OWIN app func, are sent"
+              ["After OWIN"]
+              actual
+        | false, _ -> Tests.failtest "X-Custom-After is missing"
+
+      runWithConfig composedApp |> reqResp Types.GET "/owin" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "OWIN middleware run around OWIN app, set before" <| fun _ ->
+      let noContent (env : OwinEnvironment) =
+        env.[OwinConstants.responseStatusCode] <- box 204
+        async.Return ()
+      
+      let setCustomHeaders = OwinMidFunc(fun next -> OwinAppFunc(fun env ->
+          let responseHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.requestHeaders]
+          responseHeaders.["X-Custom-Before"] <- [|"Before OWIN"|]
+          next.Invoke(env).ContinueWith(fun _ ->
+            let responseHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.requestHeaders]
+            responseHeaders.["X-Custom-After"] <- [|"After OWIN"|]
+          )
+        ))
+
+      let composedApp =
+        path "/owin"
+          // NOTE: really need some sort of splitter, as found in Freya.
+          >>= OwinApp.ofMidFunc setCustomHeaders
+          >>= OwinApp.ofApp noContent
+      
+      let asserts (result : HttpResponseMessage) =
+        eq "Http Status Code" HttpStatusCode.NoContent result.StatusCode
+        eq "Reason Phrase set by server" "No Content" result.ReasonPhrase
+
+        match result.Headers.TryGetValues("X-Custom-Before") with
+        | true, actual ->
+          eqs "Headers set before the OWIN app func, are sent"
+              ["Before OWIN"]
+              actual
+        | false, _ -> Tests.failtest "X-Custom-Before is missing"
+
+        match result.Headers.TryGetValues("X-Custom-Before") with
+        | true, actual ->
+          eqs "Headers set after the OWIN app func, are sent"
+              ["After OWIN"]
+              actual
+        | false, _ -> Tests.failtest "X-Custom-After is missing"
+
+      runWithConfig composedApp |> reqResp Types.GET "/owin" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "OWIN middleware run around OWIN app, set after" <| fun _ ->
+      let noContent (env : OwinEnvironment) =
+        env.[OwinConstants.responseStatusCode] <- box 204
+        async.Return ()
+      
+      let setCustomHeaders = OwinMidFunc(fun next -> OwinAppFunc(fun env ->
+          let responseHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.requestHeaders]
+          responseHeaders.["X-Custom-Before"] <- [|"Before OWIN"|]
+          next.Invoke(env).ContinueWith(fun _ ->
+            let responseHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.requestHeaders]
+            responseHeaders.["X-Custom-After"] <- [|"After OWIN"|]
```

Comment:
IIRC, this _should_ fail unless the middleware buffers the response itself. I need to double-check.